src/workerd/api/streams/standard.c++

```diff
@@ -2911,22 +2911,69 @@ private:
   kj::Canceler canceler;
   bool end;
 
+  // The intent with the accumulator is to attempt to hopefully improve performance
+  // a bit when the source stream is made up of a bunch of small writes. Instead of
+  // passing a bunch of individual small writes down to the sink, we'll accumulate
+  // those and pass them in a batch.
+  struct Accumulator {
+    // We will accumulate until we either queue at least THRESHOLD bytes or until
+    // we have at least MAX_COUNT pieces.
+    static constexpr size_t THRESHOLD = 4096;
+    static constexpr size_t MAX_COUNT = 16;
+    kj::Vector<kj::Array<kj::byte>> pieces;
+    size_t total = 0;
+    bool pendingClose = false;
+
+    // Adds a piece to the collection, adjusting the total buffered size accordingly.
+    void add(kj::Array<kj::byte> piece) {
+      total += piece.size();
+      pieces.add(kj::mv(piece));
+    }
+
+    kj::Array<kj::Array<kj::byte>> release() {
+      auto ret = pieces.releaseAsArray();
+      pieces.clear();
+      total = 0;
+      return kj::mv(ret);
+    }
+  };
+
   bool isErroredOrClosed() {
     return state.template is<kj::Exception>() ||
            state.template is<StreamStates::Closed>();
   }
 
+  kj::Promise<void> write(WritableStreamSink& sink,
+                          kj::Array<kj::Array<kj::byte>> pieces,
+                          bool end = false) {
+    kj::Array<const kj::ArrayPtr<const kj::byte>> views = KJ_MAP(piece, pieces) {
```

Comment:
Non-blocking, but I suppose you could potentially avoid an allocation with a `KJ_STACK_ARRAY` and explicit for loop instead of a `KJ_MAP`.